runtime/fiber.c

```diff
@@ -183,20 +186,23 @@ alloc_size_class_stack_noexc(mlsize_t wosize, int cache_bucket,
 
 /* allocate a stack with at least "wosize" usable words of stack */
 static struct stack_info* alloc_stack_noexc(mlsize_t wosize, value hval,
-                                            value hexn, value heff)
+                                            value hexn, value heff, uintnat id)
 {
   int cache_bucket = stack_cache_bucket (wosize);
-  return alloc_size_class_stack_noexc(wosize, cache_bucket, hval, hexn, heff);
+  return alloc_size_class_stack_noexc(wosize, cache_bucket, hval, hexn, heff,
+      id);
 }
 
 #ifdef NATIVE_CODE
 
 value caml_alloc_stack (value hval, value hexn, value heff) {
+  static __thread uintnat id = 1;
```

Comment:
I'm wary of depending on Max_domains remaining constant, given that I'm dreaming of making the constant go away at some point. I remember you told me that *object* unique identifiers use a cool trick to avoid excessive synchronizations by reserving intervals of size 1024 and only synchronizing on a global counter at the boundary, why not use the same trick here? (Can we even reuse the code instead of duplicating the logic?)

Reply:
The performance of fiber stack allocation is not critical unlike switching between fibers (through `perform`, `continue` and `discontinue`). We measured the performance of these operations in the PLDI 21 paper (https://arxiv.org/pdf/2104.00250.pdf, Section 6.3). We see that installing new effect handlers (which includes allocating a fiber stack) is the most expensive operation. That said, in practice, I would expect switching between fibers to outweigh the number of fiber stack allocations by a large factor. Hence, I would go for the `_Atomic int64_t` if it keeps the code simple. 

Factoring out the code for reserving intervals seems like a useful addition. I'd prefer that we do it in a separate PR.

Reply:
Sounds very reasonable to me.

Reply:
Also confirmed the same patterns post-PLDI paper in #10875, where the switch to mmap instead of the (slab allocated) malloc for fibre stack allocation made no practical performance difference.

Reply:
The "local reservation pattern" is a bit invasive, as it requires per-domain value (of course!), and thus more ceremony in Caml_state; the ceremony effectively discouraged me from writing a follow-up PR. I'm convinced by the idea that the memory allocation that follows dominates the performance cost of the atomic increment here, and we can always revisit if we learn otherwise.